### PR TITLE
[schema] Add IsOverlay option to disable codegen for particular types

### PR DESCRIPTION
**Pull request by lblackstone**
# Description

Related to https://github.com/pulumi/docs/issues/6740

Add a new `IsOverlay` option to schema types and functions that allows providers to document overlays in the schema. This makes it easier to generate API docs consistently, even for code that is generated outside of the typical codegen process.

Fixes # (issue)

## Checklist

- [ ] I have added tests that prove my fix is effective or that my feature works

- [ ] I have updated the [CHANGELOG-PENDING](https://github.com/pulumi/pulumi/blob/master/CHANGELOG_PENDING.md) file with my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Service API version

### CODE REVIEW

File: pkg/codegen/schema/schema.go [hunk 1]
```diff
@@ -1429,6 +1432,8 @@ type ResourceSpec struct {
 	IsComponent bool `json:"isComponent,omitempty" yaml:"isComponent,omitempty"`
 	// Methods maps method names to functions in this schema.
 	Methods map[string]string `json:"methods,omitempty" yaml:"methods,omitempty"`
+	// IsOverlay indicates whether the resource code is an overlay provided by the package.
+	IsOverlay bool `json:"is_overlay,omitempty" yaml:"isOverlay,omitempty"`
```
Comment (justinvp) [inline pkg/codegen/schema/schema.go:1436]:
```suggestion
	IsOverlay bool `json:"isOverlay,omitempty" yaml:"isOverlay,omitempty"`
```

?

**Comment by pgavlin** [comment]
This seems pretty reasonable to me. Feels like we might want this on all schema members (i.e. types and functions in addition to resources).

**Comment by lblackstone** [comment]
> This seems pretty reasonable to me. Feels like we might want this on all schema members (i.e. types and functions in addition to resources).

Ha, just pushed an update that expands it to types. Let me check if that also supports functions.

**Comment by emiliza** [comment]
Would be good to get this added in the dev docs as well 🙏  https://github.com/pulumi/pulumi/blob/master/developer-docs/providers/metaschema.md

**Comment by lblackstone** [comment]
I'm testing this now with the k8s provider, but open to suggestions on any unit tests that might make sense as well.

File: pkg/codegen/go/gen.go [hunk 2]
```diff
@@ -2545,6 +2545,10 @@ func (pkg *pkgContext) genResourceModule(w io.Writer) {
 		fmt.Fprintf(w, "func (m *module) Construct(ctx *pulumi.Context, name, typ, urn string) (r pulumi.Resource, err error) {\n")
 		fmt.Fprintf(w, "\tswitch typ {\n")
 		for _, r := range pkg.resources {
+			if r.IsOverlay {
```
Comment (viveklak) [inline pkg/codegen/go/gen.go:2548]: Do we need to add a test schema(s) for ensuring we don't leak overlays by mistake here and in other languages' codegen?

Comment (lblackstone) [reply pkg/codegen/go/gen.go:2548]: Yeah, probably so.

Comment (lblackstone) [reply pkg/codegen/go/gen.go:2548]: Done, and caught a bug in the process!

**Comment by lblackstone** [comment]
@viveklak @justinvp This is RFR now.

File: pkg/codegen/go/gen.go [hunk 3]
```diff
@@ -3248,7 +3290,7 @@ func GeneratePackage(tool string, pkg *schema.Package) (map[string][]byte, error
 		}
 
 		// If there are resources in this module, register the module with the runtime.
-		if len(pkg.resources) != 0 {
+		if len(pkg.resources) != 0 && !allResourcesAreOverlays(pkg.resources) {
```
Comment (viveklak) [inline pkg/codegen/go/gen.go:3293]: Nice catch!

**Review by viveklak — APPROVED** [review]

**Pull request merged**